lib/ansible/plugins/connection/ssh.py

```diff
@@ -489,11 +499,18 @@ def _build_command(self, binary, *other_args):
 
         # If we want to use password authentication, we have to set up a pipe to
         # write the password to sshpass.
-
-        if self._play_context.password:
+        pkcs11_provider = self.get_option("pkcs11_provider")
+        if self._play_context.password or len(pkcs11_provider) >= 1:
             if not self._sshpass_available():
-                raise AnsibleError("to use the 'ssh' connection type with passwords, you must install the sshpass program")
+                raise AnsibleError("to use the 'ssh' connection type with passwords or pkcs11_provider, you must install the sshpass program")
 
+        if self._play_context.password and len(pkcs11_provider) >= 1:
```

Comment:
Related to the discussion of making this inventory-configurable above: seems like using the presence of the password as a silent disable switch could be confusing if someone forgot to explicitly specify a PIN for the host/globally.

Reply:
Maybe if pkcs11 provider is set, but no password is set I could print an error message?  Should I just throw an Ansible exception or whats the best method to error on this condition to keep consistent with the rest of the Ansible source code?

Reply:
Still curious if you think this is a viable solution?